### PR TITLE
add index to node-authorizer for high cardinality vertices

### PR DESCRIPTION
**Pull request by liggitt**
follow-up to https://github.com/kubernetes/kubernetes/pull/62856#issuecomment-382788780

explores adding an index to high-cardinality vertices in the node authorizer to reduce CPU usage for high density namespaces

* first commit is a refactor only - cc @mtaufen 
* second commit adds an optional per-vertex index we can maintain when there are sufficient outgoing edges.

benchmark results:
* shared_secret_via_pod cases are ~1000x faster
* throughput on processing of graph modifications is 50% higher
* there is more variance on graph modifications requiring index updates (though the 100 index-impacting graph modifications per second might be a higher-than-realistic write rate)

data profile (5000 pods per namespace, assigned to 5000 nodes, shared service account and secret):
```
        opts := sampleDataOpts{
                // To simulate high replication in a small number of namespaces:
               nodes:       5000,
               namespaces:  10,
               podsPerNode: 10,
...
```

command:
```
$ go test ./plugin/pkg/auth/authorizer/node/  -bench Authorization  -benchmem -v 
```

before
```
BenchmarkAuthorization/allowed_node_configmap-8                                  557 ns/op   530 B/op   11 allocs/op   3000000
BenchmarkAuthorization/allowed_configmap-8                                       539 ns/op   530 B/op   11 allocs/op   3000000
BenchmarkAuthorization/allowed_secret_via_pod-8                                  605 ns/op   529 B/op   11 allocs/op   3000000
BenchmarkAuthorization/allowed_shared_secret_via_pod-8                        215974 ns/op   792 B/op   19 allocs/op      5000
BenchmarkAuthorization/disallowed_node_configmap-8                               823 ns/op   694 B/op   17 allocs/op   2000000
BenchmarkAuthorization/disallowed_configmap-8                                    888 ns/op   691 B/op   17 allocs/op   2000000
BenchmarkAuthorization/disallowed_secret_via_pod-8                               868 ns/op   694 B/op   17 allocs/op   2000000
BenchmarkAuthorization/disallowed_shared_secret_via_pvc-8                       1216 ns/op   948 B/op   22 allocs/op   1000000
BenchmarkAuthorization/disallowed_pvc-8                                          918 ns/op   691 B/op   17 allocs/op   2000000
BenchmarkAuthorization/disallowed_pv-8                                          1095 ns/op   839 B/op   19 allocs/op   2000000
BenchmarkAuthorization/disallowed_attachment_-_no_relationship-8                 867 ns/op   677 B/op   16 allocs/op   2000000
BenchmarkAuthorization/disallowed_attachment_-_feature_disabled-8                220 ns/op   208 B/op    2 allocs/op  10000000
BenchmarkAuthorization/allowed_attachment_-_feature_enabled-8                    687 ns/op   594 B/op   12 allocs/op   2000000
BenchmarkAuthorization/contentious_allowed_node_configmap-8                      656 ns/op   530 B/op   11 allocs/op   3000000
BenchmarkAuthorization/contentious_allowed_configmap-8                           659 ns/op   529 B/op   11 allocs/op   2000000
BenchmarkAuthorization/contentious_allowed_secret_via_pod-8                      654 ns/op   529 B/op   11 allocs/op   2000000
BenchmarkAuthorization/contentious_allowed_shared_secret_via_pod-8            234308 ns/op  1022 B/op   22 allocs/op      5000
BenchmarkAuthorization/contentious_disallowed_node_configmap-8                  1118 ns/op   692 B/op   17 allocs/op   1000000
BenchmarkAuthorization/contentious_disallowed_configmap-8                       1054 ns/op   692 B/op   17 allocs/op   1000000
BenchmarkAuthorization/contentious_disallowed_secret_via_pod-8                  1059 ns/op   691 B/op   17 allocs/op   2000000
BenchmarkAuthorization/contentious_disallowed_shared_secret_via_pvc-8           1403 ns/op   949 B/op   22 allocs/op   1000000
BenchmarkAuthorization/contentious_disallowed_pvc-8                             1058 ns/op   692 B/op   17 allocs/op   2000000
BenchmarkAuthorization/contentious_disallowed_pv-8                              1237 ns/op   838 B/op   19 allocs/op   1000000
BenchmarkAuthorization/contentious_disallowed_attachment_-_no_relationship-8    1022 ns/op   676 B/op   16 allocs/op   1000000
BenchmarkAuthorization/contentious_disallowed_attachment_-_feature_disabled-8    260 ns/op   209 B/op    2 allocs/op   5000000
BenchmarkAuthorization/contentious_allowed_attachment_-_feature_enabled-8        793 ns/op   594 B/op   12 allocs/op   2000000
--- BENCH: BenchmarkAuthorization
   node_authorizer_test.go:596: graph modifications during non-contention test: 0
   node_authorizer_test.go:593: graph modifications during contention test: 961
   node_authorizer_test.go:594: <1ms=774, <10ms=32, <25ms=14, <50ms=29, <100ms=62, <250ms=46, <500ms=2, <1000ms=1, >1000ms=1
```

after
```
BenchmarkAuthorization/allowed_node_configmap-8                                  629 ns/op   530 B/op   11 allocs/op   3000000
BenchmarkAuthorization/allowed_configmap-8                                       641 ns/op   530 B/op   11 allocs/op   3000000
BenchmarkAuthorization/allowed_secret_via_pod-8                                  591 ns/op   530 B/op   11 allocs/op   3000000
BenchmarkAuthorization/allowed_shared_secret_via_pod-8                           217 ns/op   160 B/op    1 allocs/op  10000000
BenchmarkAuthorization/disallowed_node_configmap-8                               912 ns/op   693 B/op   17 allocs/op   2000000
BenchmarkAuthorization/disallowed_configmap-8                                    913 ns/op   694 B/op   17 allocs/op   2000000
BenchmarkAuthorization/disallowed_secret_via_pod-8                               881 ns/op   691 B/op   17 allocs/op   2000000
BenchmarkAuthorization/disallowed_shared_secret_via_pvc-8                       1271 ns/op   952 B/op   22 allocs/op   1000000
BenchmarkAuthorization/disallowed_pvc-8                                          903 ns/op   694 B/op   17 allocs/op   2000000
BenchmarkAuthorization/disallowed_pv-8                                          1024 ns/op   836 B/op   19 allocs/op   1000000
BenchmarkAuthorization/disallowed_attachment_-_no_relationship-8                1187 ns/op   678 B/op   16 allocs/op   2000000
BenchmarkAuthorization/disallowed_attachment_-_feature_disabled-8                250 ns/op   209 B/op    2 allocs/op  10000000
BenchmarkAuthorization/allowed_attachment_-_feature_enabled-8                    694 ns/op   594 B/op   12 allocs/op   2000000
BenchmarkAuthorization/contentious_allowed_node_configmap-8                      732 ns/op   530 B/op   11 allocs/op   2000000
BenchmarkAuthorization/contentious_allowed_configmap-8                           820 ns/op   530 B/op   11 allocs/op   2000000
BenchmarkAuthorization/contentious_allowed_secret_via_pod-8                     1082 ns/op   531 B/op   11 allocs/op   1000000
BenchmarkAuthorization/contentious_allowed_shared_secret_via_pod-8               274 ns/op   160 B/op    1 allocs/op   5000000
BenchmarkAuthorization/contentious_disallowed_node_configmap-8                  1332 ns/op   693 B/op   17 allocs/op   1000000
BenchmarkAuthorization/contentious_disallowed_configmap-8                       1534 ns/op   693 B/op   17 allocs/op   1000000
BenchmarkAuthorization/contentious_disallowed_secret_via_pod-8                  1077 ns/op   692 B/op   17 allocs/op   1000000
BenchmarkAuthorization/contentious_disallowed_shared_secret_via_pvc-8           1976 ns/op   949 B/op   22 allocs/op   1000000
BenchmarkAuthorization/contentious_disallowed_pvc-8                             1297 ns/op   694 B/op   17 allocs/op   1000000
BenchmarkAuthorization/contentious_disallowed_pv-8                              1632 ns/op   837 B/op   19 allocs/op   1000000
BenchmarkAuthorization/contentious_disallowed_attachment_-_no_relationship-8    1394 ns/op   677 B/op   16 allocs/op   1000000
BenchmarkAuthorization/contentious_disallowed_attachment_-_feature_disabled-8    320 ns/op   209 B/op    2 allocs/op   5000000
BenchmarkAuthorization/contentious_allowed_attachment_-_feature_enabled-8       1055 ns/op   595 B/op   12 allocs/op   2000000
--- BENCH: BenchmarkAuthorization
    node_authorizer_test.go:629: graph modifications during non-contention test: 0
    node_authorizer_test.go:626: graph modifications during contention test: 1424
    node_authorizer_test.go:627: <1ms=0, <10ms=569, <25ms=340, <50ms=145, <100ms=101, <250ms=160, <500ms=61, <1000ms=42, >1000ms=6
```

```release-note
NONE
```

### CODE REVIEW

File: plugin/pkg/auth/authorizer/node/graph.go [hunk 1]
```diff
@@ -233,6 +258,41 @@ func (g *Graph) removeVertex_locked(v *namedVertex) {
...
+func (g *Graph) recomputeDestinationIndex_locked(n graph.Node) {
+	// don't maintain indices for nodes with few edges
+	edgeCount := g.graph.Degree(n)
+	if edgeCount < g.destinationEdgeThreshold {
+		delete(g.destinationEdgeIndex, n.ID())
+		return
+	}
+
+	// get or create the index
+	index := g.destinationEdgeIndex[n.ID()]
```
Comment (mikedanese) [inline plugin/pkg/auth/authorizer/node/graph.go]: How much does not allocating a new index every recompute help?

**Review by mikedanese — COMMENTED** [review]
Looks reasonable

File: plugin/pkg/auth/authorizer/node/graph.go (hunk 1 above)
Comment (liggitt) [reply plugin/pkg/auth/authorizer/node/graph.go]: back of envelope calculations on a 5000 node cluster with high-scale replicasets (which is really the type of scenario this index is for):

* 60k per rebuild (5000 entries, average of 13 bytes per entry - https://play.golang.org/p/aTH-j0tSPCr)
* 2 highly shared objects per pod (service account and token secret)
* 2 rebuilds per object per graph-modifying object write (on edge removal and re-add)
* ~100 graph-modifying writes per second

That gives ~20MB/sec in allocations (~1.3GB/min, ~80GB/hour)

Comment (liggitt) [reply plugin/pkg/auth/authorizer/node/graph.go]: added a benchmark to check index maintenance cost (time-wise and allocation-wise) processing a pod modification event on a 5000-node scenario:

with reuse (mark current members, sweep old members):
```
1508551 ns/op	    2034 B/op	      42 allocs/op
```

without reuse (reallocate every time we recompute the index):
```
1874243 ns/op	  271495 B/op	     282 allocs/op
```

so in this scenario, 265kb per op (probably from reallocation while growing the index from zero) vs 2kb for reuse.

interestingly, the reallocate case is also consistently slower than the mark/sweep case.

File: plugin/pkg/auth/authorizer/node/graph.go [hunk 2]
```diff
@@ -92,6 +97,9 @@ func NewGraph() *Graph {
 	return &Graph{
 		vertices: map[vertexType]namespaceVertexMapping{},
 		graph:    simple.NewDirectedAcyclicGraph(0, 0),
+
+		destinationEdgeIndex:     map[int]intSet{},
+		destinationEdgeThreshold: 100,
```
[the remark this answers is not in the document]
Comment (liggitt) [reply plugin/pkg/auth/authorizer/node/graph.go]: after more benchmarking, upped this to 200, which was the point at which iteration slowed authz checks by an order of magnitude

File: plugin/pkg/auth/authorizer/node/intset.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+// intSetGeneration represents a single generation in the set
+// the set is only capable of representing two generations
+type intSetGeneration bool
+
+// intSet maintains a set of ints, and supports promoting and culling the previous generation.
+type intSet map[int]intSetGeneration
+
+// Has returns true if the specified int is in the set
+func (s intSet) Has(i int) bool {
```
[the remark this answers is not in the document]
Comment (liggitt) [reply plugin/pkg/auth/authorizer/node/intset.go]: unexported

File: plugin/pkg/auth/authorizer/node/intset.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	_, present := s[i]
+	return present
+}
+
+// TheNextGeneration returns the marker that can be passed to includeInGeneration to promote, and limitToGeneration to cull unpromoted items
+func (s intSet) TheNextGeneration() intSetGeneration {
```
[the remark this answers is not in the document]
Comment (liggitt) [reply plugin/pkg/auth/authorizer/node/intset.go]: added documentation and made the current generation holder internal to the index

File: plugin/pkg/auth/authorizer/node/intset.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+// intSetGeneration represents a single generation in the set
+// the set is only capable of representing two generations
```
[the remark this answers is not in the document]
Comment (liggitt) [reply plugin/pkg/auth/authorizer/node/intset.go]: added docs

**Comment by liggitt** [comment]
@mikedanese comments addressed

@wojtek-t - would be interested in seeing a profile captured from a run on this PR to compare to the one mentioned in https://github.com/kubernetes/kubernetes/issues/62808#issuecomment-382651517. is there a PR-equivalent of that job we can run?

**Comment by liggitt** [comment]
rebased

**Comment by liggitt** [comment]
/hold
I wanted to add some more unit tests to verify index invalidation on update

**Comment by liggitt** [comment]
/hold cancel

added an index invalidation test

**Comment by wojtek-t** [comment]
> @wojtek-t - would be interested in seeing a profile captured from a run on this PR to compare to the one mentioned in #62808 (comment). is there a PR-equivalent of that job we can run?

Given it's already merged, we should have some results from upcoming postsubmits. I will let you know (probably on Monday when back from kubecon).

**Comment by liggitt** [comment]
looks like https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/148/ was the last run completed prior to merge

can compare http://gcsweb.k8s.io/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/148/artifacts/profiles/ to the next run

**Comment by liggitt** [comment]
Comparison of run 148 and 149

Before:
```
      "data": {
        "Perc50": 0.11,
        "Perc90": 0.163,
        "Perc99": 1.428
      },
      "unit": "ms",
      "labels": {
        "Count": "10415010",
        "Resource": "secrets",
        "Scope": "namespace",
        "Subresource": "",
        "Verb": "GET"
      }
```

After, cut 99th percentile in half:
```
      data": {
        "Perc50": 0.101,
        "Perc90": 0.153,
        "Perc99": 0.708
      },
      "unit": "ms",
      "labels": {
        "Count": "10314765",
        "Resource": "secrets",
        "Scope": "namespace",
        "Subresource": "",
        "Verb": "GET"
      }
``` 


Sample CPU profiles before:
* https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/148/artifacts/profiles/ApiserverCPUProfile_2018-05-02T09:12:19Z.pdf
* https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/148/artifacts/profiles/ApiserverCPUProfile_2018-05-02T10:07:50Z.pdf

Sample CPU profiles after:
* https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/149/artifacts/profiles/ApiserverCPUProfile_2018-05-04T09:15:14Z.pdf
* https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/149/artifacts/profiles/ApiserverCPUProfile_2018-05-04T10:10:45Z.pdf


Node authorization dropped from 4-12% before to not being present in the profile after. The index-building code path did not appear in the CPU profile either

Will keep an eye on memory use trends over the next few runs to make sure the index memory use impact wasn’t significant

**Comment by wojtek-t** [comment]
Great - thanks for checking!

**Pull request merged**